[AN] 모두의 보따리 인기 해시태그 기능 구현

moondev03 opened the pull request:
<!-- PR 제목 예시: [AN] 로그인 화면 UI 구현 / [BE] 회원가입 API 추가 / [ALL] 공통 유틸 정리 -->

## ✅ 작업 개요
<!-- 어떤 기능/수정/리팩토링인지 한 줄로 설명 (예: 로그인 화면 UI 구현) -->
- 모두의 보따리 인기 해시태그 기능 구현

<br>

## 🎯 관련 이슈
<!-- 관련된 이슈 번호를 태깅해주세요. -->
- Closed #664 

<br>

## 🔍 상세 내용
<!-- 
작업한 내용을 구체적으로 작성해주세요.
- 로그인 화면 UI 구성 (이메일, 비밀번호 입력란, 로그인 버튼)
- 입력값 유효성 검사 추가
- ViewModel 및 상태 관리 로직 연동
-->
### 개선 사항
- 기존에는 **하나의 ViewModel에서 모든 로직을 처리**했으나, 각 탭을 **독립적인 스크린**(ViewModel 단위)으로 분리하여 구조를 개선했습니다.  
- **아이템 가시성 애니메이션**을 추가하여 화면 전환 및 목록 변화 시 시각적 자연스러움을 높였습니다.  

### 인기 해시태그 영역 개선
- 인기 해시태그를 **템플릿 목록과 검색창 사이**에 배치했을 때, 상단 영역이 복잡하게 느껴졌습니다.  
- 이에 따라 **스크롤 중에는 목록에만 집중할 수 있도록**, 최상단이 아닐 경우 **검색창과 인기 해시태그 목록을 숨기도록** 처리했습니다.  


<br>

## 📸 스크린샷 (선택)
<!-- 시각적 변경사항이 있다면 스크린샷 첨부해주세요. -->
[Screen_recording_20251029_144104.webm](https://github.com/user-attachments/assets/310debcf-d4ed-4b04-99a6-f87a9b3c9c3c)



<br>

## 💬 기타 참고사항
<!-- 
코드 리뷰어가 참고해야 할 사항이 있다면 적어주세요.
- 추후 `회원가입` 화면에서도 재사용 가능한 UI 컴포넌트로 분리할 예정입니다.
- `AuthViewModel` 내 중복 로직은 이후 리팩토링에서 정리 예정입니다.
-->

<br>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **새 기능**
  * 인기 해시태그 조회 API 및 관련 화면/뷰모델·유스케이스 추가
  * 인기 해시태그 기반 검색과 페이징 연동 제공

* **개선 사항**
  * 칩 클릭 리플 피드백 및 인터랙션 추가
  * 템플릿 목록의 항목 애니메이션·클릭 반응·스크롤·FAB 동작 개선
  * 검색·페이징 결과 병합 로직 개선
  * 그림자 처리와 UI 프리뷰 개선

* **기타**
  * 북마크 토글 및 실패 알림 처리 추가
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

moondev03 committed:
refactor: 템플릿 화면 ViewModel 분리 및 구조 개선

- 기존 `TemplateViewModel`이 "전체 템플릿"과 "나의 템플릿" 화면의 모든 로직을 담당하여 비대해졌던 문제를 해결하기 위해, 각 화면에 특화된 ViewModel(`MainTemplateViewModel`, `MyTemplateViewModel`)로 분리했습니다.
- 이 과정에서 화면별로 필요한 UiState와 UiEvent를 새로 정의하고, 관련 없는 로직을 제거하여 각 ViewModel의 책임을 명확히 했습니다.
- 또한, 페이지네이션 처리 시 UseCase에서 `pageable.merge`를 직접 호출하지 않고 ViewModel에서 처리하도록 수정하여 데이터 흐름을 개선했습니다.
- `TemplateBottariScreen`을 각 탭에 해당하는 `MainTemplateScreen`, `MyTemplateScreen`으로 분리하여 컴포저블의 재사용성과 독립성을 높였습니다.

moondev03 committed:
refactor: 북마크 템플릿 아이템의 푸터 미표시 로직 변경

- 북마크 템플릿 목록의 아이템은 작성자, 담아간 횟수 등 푸터 정보가 필요하지 않습니다.
- 기존에는 `author`를 빈 문자열로, `takenCount`를 0으로 전달하여 푸터를 숨겼으나, `takenCount`를 -1로 전달하여 명시적으로 푸터를 표시하지 않도록 조건을 수정했습니다.

moondev03 committed:
feat: 템플릿 목록에 애니메이션 추가 및 스크롤에 따른 검색창 표시/숨김 기능 구현

- `LazyColumn`의 각 아이템에 `animateItem()`을 적용하여 목록에 애니메이션 효과를 추가합니다.
- 메인 템플릿 화면에서 스크롤을 내려 목록의 상단이 보이지 않을 경우 검색창이 사라지고, 다시 최상단으로 스크롤하면 검색창이 나타나도록 수정했습니다.

moondev03 committed:
feat: 북마크 상태 실시간 동기화 기능 추가

- `ObserveAllBookmarksUseCase`를 사용하여 북마크의 모든 변경사항을 관찰합니다.
- 북마크 상태가 변경될 때마다 템플릿 목록의 북마크 상태를 실시간으로 업데이트하여 화면에 반영합니다.

moondev03 committed:
feat: BottariChip에 클릭 기능 및 리플 효과 추가

- `BottariChip` 컴포저블에 `onClick` 파라미터를 추가하여 클릭 이벤트를 처리할 수 있도록 구현했습니다.
- `clickable` 모디파이어를 적용하여 칩 클릭 시 리플(ripple) 효과가 나타나도록 개선했습니다.

moondev03 committed:
feat: 인기 해시태그 조회 기능 추가

moondev03 committed:
feat: 자식의 가로 스크롤 시 부모 스크롤을 차단하는 확장 함수 추가

- `LazyColumn` 내부에 `LazyRow`와 같이 중첩된 스크롤 구조에서, 자식 컴포저블이 가로 스크롤을 시작할 때 부모의 세로 스크롤이 함께 동작하는 문제를 해결하기 위해 `rememberBlockParentAfterChild` 확장 함수를 추가했습니다.
- 이 함수는 `NestedScrollConnection`을 사용하여 자식의 가로 스크롤(`onPostScroll`) 및 플링(`onPostFling`) 이벤트 발생 시, 부모 컴포저블로 스크롤 이벤트가 전파되는 것을 차단합니다.

moondev03 committed:
refactor: NestedScrollConnection 로직을 확장 함수로 분리

- `MainTemplateScreen.kt` 내에 정의되어 있던 `rememberBlockParentAfterChild` Composable 함수를 `compose.common.extension` 패키지로 이동시켰습니다.
- 이를 통해 중첩 스크롤 시 부모의 스크롤을 막는 로직을 다른 화면에서도 재사용할 수 있도록 개선했습니다.

moondev03 committed:
refactor: 검색창 placeholder 텍스트 수정

- `MainTemplateScreen`의 검색창 placeholder 텍스트를 "제목이나 해시태그를 입력하세요"에서 "제목을 입력하거나 해시태그를 눌러보세요"로 변경하여 사용자에게 더 명확한 행동을 안내하도록 수정했습니다.

moondev03 committed:
refactor: `dropShadow` Modifier 로직 개선 및 기능 명확화

- `dropShadow` Modifier의 구현을 개선하여 더 정확하고 유연한 그림자 효과를 제공하도록 수정했습니다.
- `spread` 파라미터가 그림자 실루엣을 중심 기준으로 균등하게 확장하도록 로직을 변경했습니다.
- 번짐(blur) 효과가 없는 경우 `BlurMaskFilter`를 `null`로 설정하여 불필요한 연산을 방지합니다.
- Modifier에 KDoc 주석을 추가하여 `blur`, `offsetX/Y`, `spread` 각 파라미터의 역할을 명확히 설명하고, 부모의 클리핑(clipping)에 의해 그림자가 잘릴 수 있다는 주의사항을 명시했습니다.

moondev03 committed:
refactor: 불필요한 코드 및 파일 정리

- `SearchTemplatesByTitleUseCase`에서 사용되지 않는 `delay` import 문을 제거했습니다.
- `TemplateColumn.kt`의 코드 포맷을 정리했습니다.
- `rememberBlockParentAfterChild.kt` 파일명을 `ScrollConnectionExtensions.kt`로 변경하여 역할을 더 명확하게 나타냈습니다.

moondev03 commented on android/Bottari/presentation/src/main/java/com/bottari/presentation/compose/home/template/main/MainTemplateViewModel.kt:
```diff
@@ -0,0 +1,193 @@
+            searchTemplatesByTitleUseCase(currentState.searchWord, wordPageable)
+                .onSuccess { newPageable ->
+                    pageable = newPageable
+                    updateState {
+                        copy(templates = newPageable.contents.map(BottariTemplateUiModel::fromDomain))
+                    }
+                    if (reset) emitEvent(MainTemplateUiEvent.SearchTemplateSuccess)
+                }.onFailure { exception ->
+                    emitEvent(MainTemplateUiEvent.FetchBottariTemplatesFailure)
+                    BottariLogger.error(exception.message, exception)
+                }
+        }.invokeOnCompletion { updateState { copy(isLoading = false, isFetched = true) } }
+    }
```
SearchTemplatesByTitleUseCase에서 처리해주고 있어

cucumber99 commented on android/Bottari/presentation/src/main/java/com/bottari/presentation/compose/home/template/component/TemplateItem.kt:
```diff
@@ -64,13 +64,12 @@ fun TemplateItem(
-            if (author.isBlank() && takenCount == 0) return@BottariBox
-            Spacer(modifier = Modifier.height(hashTagSectionPadding))
-            TemplateItemFooter(author = author, takenCount = takenCount, itemCount = items.size)
+            if (author.isNotBlank() && takenCount != -1) {
```
C: UI State에서 관리할 수는 없나요?

cucumber99 commented on android/Bottari/presentation/src/main/java/com/bottari/presentation/compose/home/template/main/MainTemplateScreen.kt:
```diff
@@ -0,0 +1,246 @@
+    onChipChange: (BottariTemplateHashtagUiModel?) -> Unit,
+    modifier: Modifier = Modifier,
+) {
+    Column {
```
R: modifier를 사용하지 않고 있는 것처럼 보여요

cucumber99 commented on android/Bottari/presentation/src/main/java/com/bottari/presentation/compose/home/template/main/MainTemplateViewModel.kt:
```diff
@@ -0,0 +1,193 @@
+                        val isMarked = bookmarks.any { bookmark -> bookmark.templateId == template.id }
+                        template.copy(isMarked = isMarked)
+                    }
+                updateState { copy(templates = updated) }
```
R: 함수로 분리해도 좋을 것 같아요

cucumber99 reviewed (changes requested):
고생하셨어요~ 리뷰 확인 부탁

moondev03 committed:
refactor: `dropShadow` Modifier 로직 개선 및 프리뷰 추가

- `dropShadow` Modifier의 그림자 계산 로직을 더 명확하게 수정했습니다.
- `spread`가 적용될 때 그림자가 중앙에서 확장되지 않고 왼쪽 상단을 기준으로 확장되던 문제를 해결했습니다.
- 컴포저블의 실제 크기(`size`)에 `spread` 값을 더해 그림자의 크기를 계산하고, `offsetX`, `offsetY` 만큼 이동시켜 그림자를 그리는 방식으로 변경했습니다.
- `DropShadowPreview`를 추가하여 Modifier의 동작을 시각적으로 확인할 수 있도록 개선했습니다.

moondev03 committed:
refactor: MainTemplateScreen의 불필요한 코드 및 수정자 정리

- `MainTemplateScreen`에서 `isScrolledToTop` 변수를 중복으로 선언하는 코드를 제거했습니다.
- `TemplateHashtagSection` 컴포저블의 `Column`에 `modifier`를 적용하고, `LazyRow`의 불필요한 `modifier`를 제거하여 코드를 정리했습니다.

moondev03 committed:
refactor: TemplateColumn 구조 개선 및 리플 효과 추가

- `TemplateColumn`의 `LazyColumn` 내부 로직을 `emptyView`, `templateItems`, `loadingBlock` 등의 확장 함수로 분리하여 가독성과 구조를 개선했습니다.
- 각 템플릿 아이템을 클릭할 때 기존의 `noRippleClickable` 대신 `clickable`과 `ripple` 효과를 적용하여 사용자에게 시각적 피드백을 제공하도록 변경했습니다.
- 아이템 간 간격(`Arrangement.spacedBy`)과 좌우 패딩을 조정하여 UI를 개선했습니다.

moondev03 committed:
fix: 인기 해시태그 정렬 순서 변경

- 인기 해시태그 목록을 사용 횟수(usageCount) 기준으로 내림차순으로 정렬하도록 수정했습니다.

moondev03 committed:
refactor: MainTemplateViewModel의 로직을 작은 함수로 분리하여 가독성 개선

- `searchTemplatesByTitle`, `searchTemplatesByHashtag`의 성공 및 실패 처리를 각각 `handleLoadNextPageSuccess`, `handleLoadNextPageFailure` 함수로 분리했습니다.
- `fetchPopularHashtags`, `observeAllBookmarks`의 `onSuccess`와 `onEach` 블록 내 로직을 별도의 핸들러 함수로 추출했습니다.
- `changeMarkedById` 함수 내에서 `also`를 사용하여 상태 업데이트 로직을 간소화했습니다.
- `observeAllBookmarks` Flow에 `catch` 연산자를 추가하여 예외 발생 시 에러를 기록하도록 처리했습니다.

Leeyerin0210 commented on android/Bottari/presentation/src/main/java/com/bottari/presentation/compose/common/modifier/DropShadow.kt:
```diff
@@ -13,14 +18,15 @@ import androidx.compose.ui.graphics.Shape
 fun Modifier.dropShadow(
     shape: Shape,
-    color: Color = BottariTheme.colors.black.copy(alpha = 0.25f),
+    color: Color = Color.Black.copy(0.25f),
```
A: 보따리 테마 내부 컬러에서 기본 컬러로 바꾸신 이유가 있으신가요?

moondev03 replied:
[d46157c](https://github.com/woowacourse-teams/2025-bottari/pull/666/commits/d46157c76cc097d47277c1d806096cd6fbdbe57b)

Leeyerin0210 commented on android/Bottari/data/src/main/java/com/bottari/data/model/remote/bottari/template/BottariTemplateFetchResponse.kt:
```diff
@@ -47,8 +47,8 @@ data class BottariTemplateFetchResponse(
     ) {
-        fun toDomain(): BottariTemplateHashtag =
-            BottariTemplateHashtag(
+        fun toDomain(): HashtagDomain =
+            HashtagDomain(
```
A: Domain이라는 접미사를 사용하신 이유가 있으신가요?

moondev03 replied:
도메인 모델과 DTO의 이름이 같아 구분을 위해 도메인 모델에 alias를 부여했습니다.

Leeyerin0210 commented on android/Bottari/presentation/src/main/java/com/bottari/presentation/compose/home/template/my/MyTemplateScreen.kt:
```diff
@@ -0,0 +1,148 @@
+    val listState = rememberLazyListState()
+    val uiState = viewModel.uiState.collectAsStateWithLifecycle()
+
+    LaunchedEffect(Unit) {
```
A: LaunchedEffect 에 Key를 Unit을 준 이유가 있나요?

moondev03 replied:
`uiEvent`에 collect를 여러 번 수행할 필요가 없기 때문입니다.

Leeyerin0210 reviewed (changes requested):
수고하셨어요~

moondev03 committed:
refactor: dropShadow Modifier의 기본 색상을 테마 색상으로 변경

- `dropShadow` Modifier의 기본 그림자 색상이 `Color.Black`으로 하드코딩되어 있던 것을 `BottariTheme.colors.black`을 사용하도록 수정했습니다.

Leeyerin0210 reviewed (approved):
수고하셨어요~
